networkx/algorithms/euler.py

```diff
@@ -147,3 +158,211 @@ def eulerian_circuit(G, source=None):
+        out_degree = G.out_degree
+    else:
+        degree = G.degree
+
+    # Verify if an euler path can be found. Complexity O(n) ?
+    for v in G:
+        deg = degree(v)
+        # directed case
+        if is_directed:
+            outdeg = out_degree(v)
+            if deg != outdeg:
+                # if we have more than 2 odd nodes no Euler path exists
+                if len(check_odd) > 2:
+                    return False
+                # is odd and we append it.
+                check_odd.append(v)
+        # undirected case
+        else:
+            if deg % 2 != 0:
+                # if we have more than 2 odd nodes no Euler path exists
+                if len(check_odd) > 2:
+                    return False
+                # is odd and we append it.
+                check_odd.append(v)
+
+    if is_directed:
+
+        first = check_odd[0]
+        second = check_odd[1]
+        if G.out_degree(first) == G.in_degree(first) + 1 and \
+           G.in_degree(second) == G.out_degree(second) + 1:
+            start = second
+        elif G.out_degree(second) == G.in_degree(second) + 1 and \
+             G.in_degree(first) == G.out_degree(first) + 1:
+            start = first
+        else:
+            start = None
+
+    else:
+        start = check_odd[0]
+
+    return start
+
+
+def _find_eulerian_path(G):
+    """Returns the edges of an Eulerian path in G, (if it exists).
+
+    An Eulerian path is a path that crosses every edge in G
+    exactly once.
+
+    Parameters
+    ----------
+    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
+    -------
+    edges: generator
+        A generator that produces edges in the Eulerian path.
+
+    Raises
+    ------
+    NetworkXError: If the graph does not have an Eulerian path.
+
+    Notes
+    -----
+    Linear time algorithm, adapted from [1]_ and [3]_.
+    Information about euler paths in [2]_.
+    Code for Eulerian circuit in [3]_.
+
+    Important: In [1], euler path is in reverse order,
+    this implementation gives the path in correct order
+    as in [3]_ for eulerian_circuit. The distinction for
+    directed graph is in using the in_degree neighbors, not the out
+    ones. for undirected, it is using itemgetter(1) for get_vertex,
+    which uses the correct vertex for this order. Also, every graph
+    has an even number of odd vertices by the Handshaking Theorem [4]_.
+
+    References
+    ----------
+    .. [1] http://www.graph-magics.com/articles/euler.php
+    .. [2] http://en.wikipedia.org/wiki/Eulerian_path
+    .. [3] https://github.com/networkx/networkx/blob/master/networkx/algorithms/euler.py
+    .. [4] https://www.math.ku.edu/~jmartin/courses/math105-F11/Lectures/chapter5-part2.pdf
+
+    Examples
+    --------
+    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'),
+                      ('S', 'E')])
+    >>> list(nx.find_eulerian_path(G))
```

Comment:
All the other examples in the file use the nx. prefix. As do most of the
documentation examples.

My examples were incorrect, using the old function name. I have updated
them to use the new nx.eulerian_path() function, which is defined at the
top level of the module.

On 20 December 2015 at 02:17, Sanket Dasgupta <notifications@github.com>
wrote:

> In networkx/algorithms/euler.py
> <https://github.com/networkx/networkx/pull/1878#discussion_r48100223>:
>
> > +    ones. for undirected, it is using itemgetter(1) for get_vertex,
> > +    which uses the correct vertex for this order. Also, every graph
> > +    has an even number of odd vertices by the Handshaking Theorem [4]_.
> > +
> > +    References
> > +    ----------
> > +    .. [1] http://www.graph-magics.com/articles/euler.php
> > +    .. [2] http://en.wikipedia.org/wiki/Eulerian_path
> > +    .. [3] https://github.com/networkx/networkx/blob/master/networkx/algorithms/euler.py
> > +    .. [4] https://www.math.ku.edu/~jmartin/courses/math105-F11/Lectures/chapter5-part2.pdf
> > +
> > +    Examples
> > +    --------
> > +    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'),
> > +                      ('S', 'E')])
> > +    >>> list(nx.find_eulerian_path(G))
>
> Should be:
>
> >>> list(find_eulerian_path(G))
>
> Also you have to add your module to this __init__.py
> <https://github.com/networkx/networkx/blob/master/networkx/algorithms/__init__.py>
> so that it is available at the top level. Then this should work fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/networkx/networkx/pull/1878/files#r48100223>.
>



-- 
Humberto Ortiz-Zuazaga